### PR TITLE
[WIP] Stabilize pdata concatenate

### PR DESCRIPTION
**Pull request by JakeDern**
# Change Summary

This is part of #1926 phase 1 and re-implements the concatenate function.  I've fixed a lot of bugs, improved performance over the existing concatenate function quite a bit (mostly by simplifying implementation) and took some steps to prepare for phase 2 which will hopefully improve the interface for this module.

Major changes:
-  Moved this to the transform module, I plan to treat concatenation, splitting, and maybe also re-indexing as additional transformations on OtapArrowRecords similar to removing transport optimized encodings
- Added an (in my opinion) better `record_batch!` macro that supports dictionaries
- Many bugs fixed for schema selection (especially structs) and nullability
- Bugs fixed for boundary conditions for dictionary cardinality selection
- Turned a lot of potential panics into schema mismatch errors
- Lots of additional unit tests

Things I deferred:

- Changing the interface to concatenate to be based on OtapBatchStore deferred to phase 2 because we can't plug it into the existing code easily
- Moving reindexing into the concatenate operation deferred to phase 2 because we couldn't plug it into the existing code easily
- Some performance improvements mentioned in TODOs

Benchmark results:

| Configuration | New Implementation | Old Implementation | Speedup |
|---------------|-------------------|-------------------|---------|
| 10 inputs, 5 points | 28.18 us | 101.03 us | **3.58x** |
| 10 inputs, 50 points | 29.82 us | 110.47 us | **3.70x** |
| 100 inputs, 5 points | 246.37 us | 951.29 us | **3.86x** |
| 100 inputs, 50 points | 267.27 us | 1,020.0 us | **3.82x** |
| 1000 inputs, 5 points | 4.47 ms | 16.62 ms | **3.72x** |
| 1000 inputs, 50 points | 4.98 ms | 17.44 ms | **3.50x** |

## What issue does this PR close?

Related to #1926 

## How are these changes tested?

Many unit tests

## Are there any user-facing changes?

No.

### CODE REVIEW

File: rust/otap-dataflow/crates/pdata/src/otap/groups.rs [hunk 1]
```diff
@@ -996,7 +996,8 @@ fn size_over_limit(max_items: Option<NonZeroU64>, size: usize) -> bool {
         .unwrap_or(false)
 }
 
-fn generic_schemaless_concatenate<const N: usize>(
+/// FIXME: pub
```
Comment (JakeDern) [inline rust/otap-dataflow/crates/pdata/src/otap/groups.rs]: Will rip this out before merge along with the whole old implementation. Just doing it last so I can keep running benchmarks if there are more changes

File: rust/otap-dataflow/crates/pdata/benches/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+### Concatenation
+```bash
+cargo bench --bench concatenate
+```
+
+## Benchmark Results
+
+The following results compare the performance of the new `concatenate` implementation versus the old `generic_schemaless_concatenate` implementation.
+
+### Single Metric Type (Gauges)
```
Comment (JakeDern) [inline rust/otap-dataflow/crates/pdata/benches/README.md]: Also going to rip this out before merge.

File: rust/otap-dataflow/crates/pdata/benches/concatenate.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+                        // Clone the same pre-generated data for each iteration
+                        data.clone()
+                    },
+                    |mut batches| {
+                        generic_schemaless_concatenate::<{ Metrics::COUNT }>(&mut batches)
+                            .expect("Concatenation failed")
+                    },
+                    BatchSize::SmallInput,
+                )
+            },
```
Comment (JakeDern) [inline rust/otap-dataflow/crates/pdata/benches/concatenate.rs]: Will remove this and everything else related to the old implementation prior to merge

File: rust/otap-dataflow/crates/pdata/benches/concatenate.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
+// Copyright The OpenTelemetry Authors
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/pdata/benches/concatenate.rs:1]: most of our other benchmarks are in `otap-dataflow/benchmarks`
https://github.com/open-telemetry/otel-arrow/tree/main/rust/otap-dataflow/benchmarks

Maybe we should move this benchmark to that location as well?

Comment (JakeDern) [reply rust/otap-dataflow/crates/pdata/benches/concatenate.rs:1]: Oh sure, I can move them if that's the convention. Just curious, is there a reason we don't keep benchmarks which are local to some crate in the crate itself? I wonder if the cargo toml there will get kind of crowded over time since it's sort of all one namespace for benchmark names

Comment (albertlockett) [reply rust/otap-dataflow/crates/pdata/benches/concatenate.rs:1]: I actually don't think there's a good reason beyond convention. Upon further inspection, it turns out I have lied to you! both quiver and query-engine have benchmarks internal to those crates, so I'm fine with keeping this here!

In the future, maybe we should move the other benchmarks that are specific to pdata into this location. Thinking about filter, attribute_transform, and transport_optimize.

File: rust/otap-dataflow/crates/pdata/src/testing/record_batch.rs [hunk 5]
```diff
@@ -0,0 +1,531 @@
...
+                .is_some()
+        );
+    }
+
+    #[test]
+    fn test_dict() {
+        use arrow::array::DictionaryArray;
+        use arrow::datatypes::Int32Type;
+        let batch =
+            record_batch!(("a", (Int32, Utf8), ([0, 1, 2], ["foo", "bar", "baz"]))).unwrap();
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/pdata/src/testing/record_batch.rs:233]: this is awesome!

Out of curiosity, does it support nullable dict keys? e.g. could I do:
```rs
record_batch!(("a", (UInt8, Utf8), ([None, Some(1), Some(2)], ["foo", "bar", "baz"]))).unwrap();
```

If so, it might be nice to have a test for that

File: rust/otap-dataflow/crates/pdata/src/error.rs [hunk 6]
```diff
@@ -34,6 +34,18 @@ pub enum Error {
...
+        "Column `{}` value type mismatch, expect: {}, actual: {}",
+        name,
+        expect,
+        actual
+    )]
+    ColumnValueTypeMismatch {
+        name: String,
+        expect: DataType,
+        actual: DataType,
+    },
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/pdata/src/error.rs:47]: nit: Is this only used for Dictionary columns? If so, would it be worthwhile to add the keyword dictionary just so folks don't misinterpret value to mean like, the actual value w/in some array was invalid versus the actual meaning which is that the array type was invalid?
```suggestion
    #[error(
        "Column `{}` value type mismatch, expect: {}, actual: {}",
        name,
        expect,
        actual
    )]
    ColumnDictionaryValueTypeMismatch {
        name: String,
        expect: DataType,
        actual: DataType,
    },
```

File: rust/otap-dataflow/crates/pdata/src/otap/transform/concatenate.rs [hunk 7]
```diff
@@ -0,0 +1,2351 @@
...
+///
+/// # Future optimizations
+///
+/// - TODO: Re-indexing probably should not be a separate operation we should decide
+/// within this function whether or not to do it and ensure it happens if required.
+/// This is deferred until we totally remove the old implementation in groups.rs
+/// due to interface incompatibility.
+///
+/// - TODO: Consider using new_unchecked for record batch construction if we're
+/// confident in it. We mostly unwerap those opterations a lot, so skipping the
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/pdata/src/otap/transform/concatenate.rs]:
```suggestion
/// confident in it. We mostly unwrap those operations a lot, so skipping the
```

File: rust/otap-dataflow/crates/pdata/src/otap/transform/concatenate.rs [hunk 8]
```diff
@@ -0,0 +1,2351 @@
...
+        for payload in select_all_mut(&mut items, i) {
+            let Some(rb) = payload.take() else {
+                continue;
+            };
+
+            let (curr_schema, columns, num_rows) = rb.into_parts();
+            let converted_columns =
+                convert(columns, num_rows, &curr_schema.fields, &new_schema.fields)?;
+            let converted = RecordBatch::try_new(new_schema.clone(), converted_columns)
+                .expect("Valid construction");
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/pdata/src/otap/transform/concatenate.rs:110]: Usually we'd put a `safety: ` comment when using `expect` (outside of benchmarks/unit tests). Something like:
```suggestion
            // safety: convert will ensure all our columns to have a type that matches the schema,
            // and doesn't modify the length of the columns, so it is safe to expect try_new to succeed
            let converted = RecordBatch::try_new(new_schema.clone(), converted_columns)
                .expect("Valid construction");
```

File: rust/otap-dataflow/crates/pdata/src/otap/transform/concatenate.rs [hunk 9]
```diff
@@ -0,0 +1,2351 @@
...
+        result[i] = Some(batch);
+    }
+
+    Ok(result)
+}
+
+// Convert the columns from one schema to another. The arguments deal in in
+// fields and columns rather than schemas and record batches so that this
+// code can work with either struct arrays or record batches.
+fn convert(
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/pdata/src/otap/transform/concatenate.rs]: nit: `///` for function docs
```suggestion
/// Convert the columns from one schema to another. The arguments deal in
/// fields and columns rather than schemas and record batches so that this
/// code can work with either struct arrays or record batches.
fn convert(
```

File: rust/otap-dataflow/crates/pdata/src/otap/transform/concatenate.rs [hunk 10]
```diff
@@ -0,0 +1,2351 @@
...
+                        // ref, so we cannot downcast_mut. Since we can't downcast_mut, we can't
+                        // break into parts. The clone isn't that bad for now since it's only a
+                        // Vec<ArrayRef>.
+                        let struct_array = struct_array
+                            .as_any()
+                            .downcast_ref::<StructArray>()
+                            .unwrap()
+                            .clone();
+
+                        let (struct_fields, struct_columns, nulls) = struct_array.into_parts();
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/pdata/src/otap/transform/concatenate.rs:159]: Would be nice if we had an `as_parts` method on arrow-rs. This could return like `(&[Fields, &[ArrayRef], Option<&NullBuffer>)`. Could be a good addition :)

File: rust/otap-dataflow/crates/pdata/src/otap/transform/concatenate.rs [hunk 11]
```diff
@@ -0,0 +1,2351 @@
...
+            }
+        }
+    }
+
+    Ok(new_columns)
+}
+
+// Select a unified schema that will satisfy all fields in the
+// RecordIndex.
+fn select_schema<'a>(index: &'a RecordIndex<'a>) -> Result<Schema> {
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/pdata/src/otap/transform/concatenate.rs]: nit: triple / for the function docs
```suggestion
/// Select a unified schema that will satisfy all fields in the RecordIndex.
fn select_schema<'a>(index: &'a RecordIndex<'a>) -> Result<Schema> {
```

File: rust/otap-dataflow/crates/pdata/src/otap/transform/concatenate.rs [hunk 12]
```diff
@@ -0,0 +1,2351 @@
...
+        index.row_count += rb.num_rows();
+
+        let fields = rb.schema_ref().fields();
+        let iter = fields.iter().zip(rb.columns());
+        index_fields(&mut index.fields, iter, None)?;
+    }
+
+    // We need a final pass to see if any fields were not present in any batch
+    // and similarly for structs to see if any struct fields were missing.
+    // When we coerce to the same schema, we have to appent nulls for missing fields.
```
Comment (albertlockett) [inline rust/otap-dataflow/crates/pdata/src/otap/transform/concatenate.rs]:
```suggestion
    // When we coerce to the same schema, we have to append nulls for missing fields.
```

File: rust/otap-dataflow/crates/pdata/src/otap/groups.rs [hunk 13]
```diff
@@ -1356,1798 +1277,3 @@ impl<T> IDRange<T> {
...
-            );
-            assert_eq!(
-                result.column(i).as_ref(),
-                expected.column(i).as_ref(),
-                "Column {} data mismatch",
-                i
-            );
-        }
-    }
-}
```
Comment (jmacd) [inline rust/otap-dataflow/crates/pdata/src/otap/groups.rs:3153]: 😅 Nice

**Review by jmacd — APPROVED** [review]
This code is very readable! Thank you.

File: rust/otap-dataflow/crates/pdata/src/testing/record_batch.rs (hunk 5 above)
Comment (JakeDern) [reply rust/otap-dataflow/crates/pdata/src/testing/record_batch.rs:233]: Currently no, but it sounds like a good addition. I'll add some TODOs!
